schema.org markup for video objects

lwrubel opened the pull request:
Resolves #824. Adds schema.org JSON for a videos that are world-downloadable. Example (https://sul-purl-stage.stanford.edu/vx195sw6395):

```<script type="application/ld+json">{"@context":"http://schema.org","@type":"VideoObject","name":"Structural relief and fish abundance","description":"In BIOHOPK185 (Ecology and Conservation of Kelp Forest Communities), students learned how to use underwater video as a tool for underwater research. In addition, students were encouraged to use the cameras to collect videos that captured the beauty of the kelp forest and the process of scientific diving at Hopkins Marine Station. In 2023, students were tasked with presenting the results of one-week research projects that used transect video surveys in a five minute movie. This collection contains these movies.\n\nFunding for video equipment was provided by the Stanford Accelerator for Learning through the Gordon and Betty Moore Foundation Grant GBMF10266 to support the work of Virtual Field Trips.","thumbnailUrl":"https://sul-stacks-stage.stanford.edu/file/druid:vx195sw6395/pf804qh8129_kfe_project_2023_mattioli_thumb.jp2","uploadDate":"2023","embedUrl":"https://embed-stage.stanford.edu/iframe/?url=https%3A%2F%2Fsul-purl-stage.stanford.edu%2Fvx195sw6395"}</script>```

Example 2: (https://sul-purl-uat.stanford.edu/bc169zr6817) checked on [Google's Rich Results Test](https://search.google.com/test/rich-results):
```<script type="application/ld+json">{"@context":"http://schema.org","@type":"VideoObject","name":"SCRF session - 1 - PIP-II","thumbnailUrl":"https://stacks-uat.stanford.edu/file/druid:bc169zr6817/bc169zr6817_SRF_AWLC_Oct20-3_thumb.jp2","uploadDate":"2020-10-20","embedUrl":"https://embed-uat.stanford.edu/iframe/?url=https%3A%2F%2Fsul-purl-uat.stanford.edu%2Fbc169zr6817"}</script>```

lwrubel commented:
@arcadiafalcone would you take a look at the specs (especially spec/lib/metadata/schema_dot_org_spec.rb) to see if the Cocina is realistic? If there are other scenarios to include in the test, let me know.

arcadiafalcone commented:
Is there a schema.org reason for joining title and subtitle with a new line rather than the more usual semicolon? https://github.com/sul-dlss/purl/blob/a94ffa50c9fa8a5e897f27d97a83e2bce3ffb92e/spec/lib/metadata/schema_dot_org_spec.rb#L93

The DOI might be in identifier.uri instead of identifier.value for some objects. https://github.com/sul-dlss/purl/blob/a94ffa50c9fa8a5e897f27d97a83e2bce3ffb92e/spec/lib/metadata/schema_dot_org_spec.rb#L199

And the ORCID may also be in either identifier.value or identifier.uri - MODS doesn't make the distinction. https://github.com/sul-dlss/purl/blob/a94ffa50c9fa8a5e897f27d97a83e2bce3ffb92e/spec/lib/metadata/schema_dot_org_spec.rb#L389

lwrubel commented:
Thanks @arcadiafalcone. I'm fixing title and subtitle concatenation since I was erroneously using the same approach as for the description. Will use semi-colon to concatenate. 

For the DOI in the identifier.uri, would this spec suffice?
https://github.com/sul-dlss/purl/blob/6eb54128e610a14b9f0d215aab589e468b20c90d/spec/lib/metadata/schema_dot_org_spec.rb#L198

For the ORCID, the current code / specs handle a contributor having `identifier.uri` or the contributor having `identifier.value` with `type` being `"orcid"`. Is there also a case where it could be `identifier.value` without a `type`?

arcadiafalcone commented:
My error, it should be title: subtitle (colon rather than semicolon).

arcadiafalcone commented:
Re: DOI in identifier.uri, so long as it confirms that it is the right domain for a DOI (not all values in this field will be DOIs).

Re: ORCID, identifier.value should always have a type. (If it doesn't, that's a data problem.)

lwrubel commented:
Thanks, @arcadiafalcone. I've adjusted the specs and code for these.

justinlittman commented on lib/metadata/schema_dot_org.rb:
```diff
@@ -44,38 +42,58 @@ def dataset?
     end
 
-    def access
+    def access?
```
Shouldn't this take into account the file permissions as well?

lwrubel replied:
I think I need help understanding if it's possible to have an object with `access.download == "world"` and the video itself is not world-downloadable. Is that the scenario you're thinking of, @justinlittman?

justinlittman replied:
Correct.

andrewjbtw replied:
It's possible. It's less likely than the inverse case, where the `access.download` is `none` but one or more files are world-downloadable.

lwrubel replied:
I've adjusted the code to look at the `access.download` permissions for the first file where `hasMimeType` includes `video`.

justinlittman commented:
@lwrubel Can a video object have more than one video file?

lwrubel commented on lib/metadata/schema_dot_org.rb:
```diff
@@ -44,38 +42,58 @@ def dataset?
 
+    def video?
+      # Only return video metadata if world-downloadable.
+      video = JsonPath.new("$.description.form[?(@['value'] == 'moving image' && @['type'] == 'resource type')]").on(@cocina_json)
```
That was @arcadiafalcone's recommendation: https://github.com/sul-dlss/purl/issues/824#issuecomment-1821005706. Is there a different field in the cocina we should consider?

lwrubel replied:
I would expect `https://cocina.sul.stanford.edu/models/media` to include audio as well, and we only want video. Am I not understanding the type correctly?

justinlittman replied:
I'm fairly sure that I'm guessing here. Alternatively, resource's type = "https://cocina.sul.stanford.edu/models/resources/video"

andrewjbtw replied:
Yes, at the resource/fileset level, media also includes audio. And image and file, for good measure.

lwrubel replied:
@arcadiafalcone could you make a recommendation about using a fileset resource type vs the descriptive metadata for videos?

arcadiafalcone replied:
Descriptive metadata is dependent on the user to create it, so the fileset resource type is probably a bit more reliable (it's a core metadata field that's usually provided, but a user could make an error).

lwrubel replied:
Thanks, that's helpful!

andrewjbtw replied:
Is the schema.org expectation that a "video" = streaming video? Or would it include a video file that may be downloadable but isn't presented in a player? If the expectation is streaming, then the fileset resource type is the only guarantee that it will be presented that way.

We may also have description that says something is a video when the video itself hasn't been digitized. I'm not aware of examples but we do have "audio" where only the record/album cover/liner notes have been digitized.

lwrubel replied:
I've updated the code to use the fileset resource type for determining it's a video.

lwrubel commented:
@justinlittman Looking at the canonical examples, I see that it is possible to have multiple video files (e.g. https://purl.stanford.edu/yj807zw8315). I'm trying to figure out what makes the most sense for schema.org. 

There is an [ItemList](https://developers.google.com/search/docs/appearance/structured-data/video#item-list) with limited usage, but it's intended for video carousels more than multi-file objects. Honestly, I'm not sure if it's important to surface thumbnails or embedUrls of more than one video, if on the PURL we currently show the first one, with access to the rest. They share descriptive metadata. (Not sure if there is rights variability among a list of videos, but we could pull the embedURL and thumbnail for the first video that is world-downloadable.)

lwrubel commented:
Noting here that there seem to very few videos that meet our criteria for the uploadDate of the object having an `event.date.type == 'publication'`. Next week I will run a report to see how few they really are. I don't know if there are other date types that might be logical for this purpose, @arcadiafalcone. Here's Google's description of the field: https://developers.google.com/search/docs/appearance/structured-data/video#video-object.

lwrubel commented:
I think I've addressed the comments, and this is otherwise fine to proceed on any further code review. We can tweak the date logic or broaden/narrow criteria further if needed, of course.

justinlittman commented on lib/metadata/schema_dot_org.rb:
```diff
@@ -13,27 +14,24 @@ def initialize(cocina_json)
 
     def schema_type?
-      dataset?
+      dataset? || video?
```
This method is poorly named since it is taking into account rights as well as the schema type.

lwrubel replied:
Changed to `render_video_metadata?`

justinlittman commented on lib/metadata/schema_dot_org.rb:
```diff
@@ -44,44 +42,74 @@ def dataset?
+      # need to find the file that is the one for the video (based on mime-type). Then get the access and download rights for that.
+      file_access = JsonPath.new('$[*].structural.contains[*][?(@.hasMimeType =~ /video/)].access.download').first(video)
+
+      return true if file_access == 'world'
```
L108-110 could just be `file_access == 'world'`

lwrubel replied:
Oh, right. Done here and in `dataset?`

justinlittman commented on lib/metadata/schema_dot_org.rb:
```diff
@@ -44,44 +42,74 @@ def dataset?
+    def format_specific_fields
+      if dataset?
+        return { "identifier": identifier,
+                 "isAccessibleForFree": access?,
```
But why publicize something that we can't provide?

lwrubel replied:
@amyehodge I am making the assumption that there are SDR-deposited datasets that are not freely available but possible to use, once the user goes through some process (e.g. a DUA or contacting the author). Or that creators of Stanford-only datasets may still want some kind of visibility in Google Dataset Search. But maybe that is not a realistic scenario or something we support?

amyehodge replied:
Another example here https://purl.stanford.edu/gh587bx9720 where the data files are actually on the PURL.

lwrubel commented on lib/metadata/schema_dot_org.rb:
```diff
@@ -44,44 +42,74 @@ def dataset?
+    def video?
+      # Only return video metadata if world-downloadable.
+      video = JsonPath.new("$.structural.contains[?(@['type'] == 'https://cocina.sul.stanford.edu/models/resources/video')]").on(@cocina_json)
+      return true if video.any? && access? && video_access?
```
Changing `access?` to `object_access?`. Changing `video?` to `render_video_metadata?`
